superset/config.py

```diff
@@ -432,6 +432,11 @@ class D3Format(TypedDict, total=False):
     # expanded into extra columns and/or arrays. This is experimental,
     # and doesn't work with all nested types.
     "PRESTO_EXPAND_DATA": False,
+    # When this feature is enabled, nested ROW types in Trino will be fully
+    # described in SQL Lab when retrieving schemas and performing previews.
+    # Unlike PRESTO_EXPAND_DATA this will not attempt to expand ARRAY or MAP
+    # types as these can't be queried with a dotted.path.syntax like ROWs
+    "TRINO_EXPAND_ROWS": False,
```

Comment:
Rather than this being a feature flag (specific to Trino) I wonder if this would make sense if this was a configuration at the database level. There's an `extra` column which has the ability to encode these types of configuration.

Reply:
Makes sense, I didn't look into opening that can of worms as I feared it'd involve db migrations and form changes but if it's being encoded via `extra` it doesn't sound too hard, I'll take a look.

Reply:
@giftig I think the [metadata_params](https://github.com/apache/superset/blob/1e37f0b41782d35d68f657bfa87aeb9055d1e6e7/superset/models/core.py#L157) subfield might be the best place to add this as it will allow this to be configurable via the UI. Otherwise one one would need to manage this via a SQL UPDATE statement.

Reply:
There's a small wrinkle here in that the `get_columns` method doesn't have a reference to the specific database, it's just a classmethod on the db spec. I could change the signature to take one; it's invoked from `Database.get_columns` anyway so there'd definitely be one available there to pass in, but then this will change the signature of a core API on the base spec, which is unfortunate.

I can also just add some extra options to the base signature and pass the flag in directly rather than a reference to the database, but then the `Database` object is forced to care about a trino-specific option, which seems incorrect. If I phrase the config as something more generic like "format options" and take a dict then that might be ok, I'll give that a try.

I actually wonder why we don't an instance of the `EngineSpec` with the `Database` and some other context present, rather than using static class methods, but I assume there was a reason for that original choice.

Reply:
I've made the change but I actually get a frontend error at the point of finishing writing a JSON object if I try to configure this in the UI:

![screenshot_2023-11-01_11-02-09_f87cc3ff](https://github.com/apache/superset/assets/2862039/d8432f6f-fd86-4a29-aa8d-4a903814c0a8)

which blanks out the page entirely as soon as I hit the closing brace. I'll raise a bug if we're not already tracking this. It exhibits very strange behaviour actually, e.g. if I type `{}` it wordlessly empties the box.

Reply:
Ok so I can't put it into `metadata_params`; that's not just arbitrary database metadata, it's actually meant to represent the sqlalchemy `MetaData` class and is being extracted out like this:

```python
from sqlalchemy import MetaData
MetaData(**extra.get("metadata_params", {}))
```
so if I add something arbitrary to it I just get:

```python
TypeError: __init__() got an unexpected keyword argument 'schema_options'
```

Reply:
It doesn't look like there's anywhere else I can put this that can be edited from the UI without making form changes. I'll have a look and see if I can use a new top-level field within `extra` by updating the form, but I assume making form changes will necessitate frontend changes, new wording, translations, and seems like a whole can of worms.

So I'd propose we just go back to the feature flagged version for simplicity for the moment, but open to further suggestions.

Reply:
@giftig there are [existing patterns](https://github.com/apache/superset/blob/a3686459a9ac4a4b916c111971de85800c0e7dee/superset/models/core.py#L216-L237) where we store configurations at the root level of the `dbs.extra` column. See [here](https://github.com/apache/superset/blob/a3686459a9ac4a4b916c111971de85800c0e7dee/superset/views/database/mixins.py#L123-L152) as well.

I speculate this information was exposed when the database connection editor modal was rewritten from the original Flask-AppBuilder view.

Reply:
I see the options you've pointed to there are some of the checkboxes exposed on the form, yeah: https://github.com/apache/superset/blob/master/superset-frontend/src/features/databases/DatabaseModal/ExtraOptions.tsx#L192

I can update this form, I'm just wondering if there's any particular approval process for copy, and how the translations get filled out. I'll add some sensible label for now and leave it for review. I may even be able to show this option only when configuring a trino db, let me see if that's relatively simple.

Reply:
Done.